src/container/index.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  NodeContainer,
+  ContainerStorage,
+  encodeUser,
+  encodeIdentity,
+  decodeUser,
+  decodeIdentity
+} from '@skygear/node-client';
+import { User, Identity } from '@skygear/node-client';
+
+import { ControllerContainer } from './ControllerContainer';
+import { save, load, ConfigDomain } from '../config';
+import { GlobalConfig } from '../types';
+
+class CLIYAMLContainerStorage implements ContainerStorage {
+  async setUser(namespace: string, user: User): Promise<void> {
+    const e = encodeUser(user);
+    const globalConfig = load(ConfigDomain.GlobalDomain) as GlobalConfig;
+    const newConfig = { ...globalConfig };
+    newConfig.user[namespace] = {
+      ...globalConfig.user[namespace],
+      user: e
+    };
```

Comment:
This statement mutates the `globalConfig.user` because `{ ...obj }` is a shadow copy operation. The following pattern allows us update deeply nested object without mutation.

```typescript
const newConfig = {
  ...globalConfig,
  user: {
    ...globalConfig.user,
    [namespace]: {
      ...globalConfig.user[namespace],
      user: e,
    }
  }
};
```

However, we always `load` the config from file system. Therefore it seems that it is simpler to do destructive update instead of bothering with the shadow copying spread operator.

```typescript
const globalConfig = load(ConfigDomain.GlobalDomain) as GlobalConfig;
globalConfig.user[namespace] = {
  ...globalConfig.user[namespace],
  user: e,
};
```

The comment should apply to most methods in this class.

Reply:
Ya, you are right. Missed it, thanks for remaining!